compositor_pipeline/src/pipeline/output/whip/establish_peer_connection.rs

```diff
@@ -0,0 +1,241 @@
+        move |connection_state: RTCIceConnectionState| {
+            debug!("Connection State has changed {connection_state}.");
+            if connection_state == RTCIceConnectionState::Connected {
+                debug!("Ice connected.");
+            } else if connection_state == RTCIceConnectionState::Failed {
+                debug!("Ice connection failed.");
+                whip_ctx
+                    .should_close
+                    .store(true, std::sync::atomic::Ordering::Relaxed);
+            }
+            Box::pin(async {})
+        },
+    ));
+
+    if let Some(keyframe_sender) = whip_ctx.request_keyframe_sender {
+        let senders = peer_connection.get_senders().await;
+        for sender in senders {
+            let keyframe_sender_clone = keyframe_sender.clone();
+            whip_ctx.tokio_rt.spawn(async move {
+                loop {
+                    if let Ok((packets, _)) = &sender.read_rtcp().await {
+                        for packet in packets {
+                            if packet
+                                .as_any()
+                                .downcast_ref::<PictureLossIndication>()
+                                .is_some()
+                            {
+                                if let Err(err) = keyframe_sender_clone.send(()) {
+                                    debug!(%err, "Failed to send keyframe request to the encoder.");
+                                };
+                            }
+                        }
+                    } else {
+                        debug!("Failed to read RTCP packets from the sender.");
+                    }
+                }
+            });
+        }
+    }
+
+    let offer = peer_connection
+        .create_offer(None)
+        .await
+        .map_err(WhipError::OfferCreationError)?;
+
+    let endpoint_url = Url::parse(&whip_ctx.options.endpoint_url)
+        .map_err(|e| WhipError::InvalidEndpointUrl(e, whip_ctx.options.endpoint_url.clone()))?;
+
+    info!("Endpoint url: {}", endpoint_url);
+
+    let mut header_map = HeaderMap::new();
+    header_map.append("Content-Type", HeaderValue::from_static("application/sdp"));
+
+    if let Some(token) = &whip_ctx.options.bearer_token {
+        let header_value_str: HeaderValue = match format!("Bearer {token}").parse() {
+            Ok(val) => val,
+            Err(err) => {
+                error!("Ivalid header token, couldn't parse: {}", err);
+                HeaderValue::from_static("Bearer")
+            }
+        };
+        header_map.append("Authorization", header_value_str);
+    }
+
+    let response = client
+        .post(endpoint_url.clone())
+        .headers(header_map)
+        .body(offer.sdp.clone())
+        .send()
+        .await
+        .map_err(|_| WhipError::RequestFailed(Method::POST, endpoint_url.clone()))?;
+
+    if let Err(err) = &response.error_for_status_ref() {
+        let status = err.status().unwrap();
```

Comment:
check status instead of error_for_status_ref. See implementatation of error_for_status_ref how they decide whether it is error or not